Export-XLF function for ADS

smartguest opened the pull request:
This PR adds a new gulp task that generates XLF files for ADS extensions, (It also adds a json file containing a list of extensions recognized as part of ADS and extensions that are from vscode) this compiles the relevant extensions that are not compiled as part of the vscode-translations-export function in ADS and exports their string resources into an external folder, This is meant for internal use, so that whenever changes are made to any extension file strings, this tool will update the XLF resource files appropriately when run manually. (Otherwise the XLF files for some of these extensions will have to be updated manually)

Charles-Gagnon commented on build/lib/locFunc.ts:
```diff
@@ -24,33 +29,141 @@ export function packageLangpacksStream(): NodeJS.ReadWriteStream {
-			es.readArray(files).pipe(result);
-		})
-		.catch(err => result.emit('error', err));
+export function createXlfFilesForExtensions(): ThroughStream {
```
It looks like this exists already in https://github.com/microsoft/azuredatastudio/blob/main/build/lib/i18n.ts - is there a reason you're duplicating it here?

smartguest replied:
I will export that instead.

smartguest commented on build/lib/locFunc.ts:
```diff
@@ -24,33 +29,141 @@ export function packageLangpacksStream(): NodeJS.ReadWriteStream {
+							return `Unknown message for key: ${key}`;
+						}
+					});
+					getXlf().addFile(`extensions/${extensionName}/package`, keys, messages);
```
The current vscode localization process is not known to me, but this is the function for creating XLFs from extensions when running vscode-translations-export

smartguest commented on i18n/ADSExtensions.json:
```diff
@@ -0,0 +1,58 @@
+		"schema-compare": "Microsoft.schema-compare",
+		"Microsoft.sqlservernotebook": "Microsoft.sqlservernotebook"
+	},
+	"VSCODEExtensions": [
```
This is a leftover from the code for updating the lang packs themselves, that step may be implemented in another PR.

smartguest commented on build/gulpfile.extensions.js:
```diff
@@ -189,6 +190,17 @@ gulp.task(task.define('extensions-ci', task.series(compileExtensionsBuildTask)))
+// {{SQL CARBON EDIT}} Need to handle localization unlike above.
+const compileLocalizationExtensionsBuildTask = task.define('compile-localization-extensions-build', task.series(
+	task.define('clean-localization-build', util.rimraf('.locbuild')),
+	task.define('bundle-marketplace-extensions-build', () => ext.packageMarketplaceExtensionsStream(false).pipe(gulp.dest('.locbuild'))),
```
Originally this was to be bundled with the task for updating the langpacks directly, and the id for the extensions folder was "vscode-extensions" instead of "extensions" when it appeared in the output folder. Since that step is to be done at a later time and there are ways to handle the different name. I will use the build folder directly, but note that it is not to be used for compilation, but rather for generating XLFs exclusively.

Charles-Gagnon commented on build/gulpfile.vscode.js:
```diff
@@ -454,6 +454,22 @@ gulp.task(task.define(
 	)
 ));
 
+// {{SQL CARBON EDIT}} Localization gulp task, similar to vscode-translations-export but only for ADS only extensions.
```
Is there some reason we couldn't use that task instead? Not sure why we need a special one for ADS-only extensions if this is actually doing basically the same thing

smartguest replied:
The vscode-translations-export task only includes extensions that are not marked as external extensions rather than all extensions due to using packagelocalextensionstream in extensions.ts. This factors into the extensions that will be bundled with ADS or packaged separately. Since some of these extensions are not meant to be bundled with ADS, either a flag passed in to indicate the usage of the packagelocalextensionstream code or creating a new function based on it is justified.

Charles-Gagnon replied:
Ok, so I would instead clarify that this only packages EXTERNAL extensions. It's technically true that those all happen to be ADS extensions, but in this case I think it makes it a lot clearer. 

And then in that case can you not use https://github.com/microsoft/azuredatastudio/blob/main/build/lib/extensions.ts#L242? It'd be nice to not have yet another list of extensions we have to remember to update if we add a new extension.

smartguest replied:
To clarify: it does not localize all of the External extensions as some of them do not have localization yet, there's also extensions that are bundled with ADS such as resource-deployment that aren't in vscode and therefore would need to be localized. This list of extensions in the code that I added therefore contains a mixture of bundled and external extensions that we currently are translating for, and that are not already translated in vscode. Only some extensions from the bundled and external extensions are used, and not all of them, hence the new list.

Charles-Gagnon replied:
Couldn't we either : 

1. Try to localize all extensions, and if it doesn't have any strings to localize then the output would just be empty (which shouldn't really impact anything)
2. Have some other way to detect if an extension should be localized, such as searching for a file that only exists if an extension should be localized (.i18n.json or package.nls.json?)

smartguest replied:
Running this on all extensions causes a too many open files message when it happens though, I'll try to add a check for package.nls.json though.

Charles-Gagnon reviewed (approved):
I don't really have the time or context to go through the logic in great deal unfortunately but nothing stands out as being a huge issue.

The main thing I'm wondering though is why we're treating the ADS extensions specially? Shouldn't we be localizing all extensions? Even the core ones brought in from VS Code? Because we still need those features localized if we're using them.

smartguest commented:
I could theoretically localize all of the extensions that are included in ADS including vscode ones but we already have them localized in the vscode repository for the most part with little to no changes. If we are willing to localize them ourselves again, I could add them to be generated as part of our xlf files along with vscode core strings as well.

Charles-Gagnon commented:
That's an interesting question, how are you planning on handling the translations for the VS Code extensions? They'd have to be bundled in with our langpacks right?

I'm not sure we want to localize those extensions again - mostly just to save the loc team from having to duplicate these strings. But it would be the safest if that's not an issue since otherwise we risk getting a version of the translations from a version of VS Code that we're not synced to.

smartguest commented:
@Charles-Gagnon with the last langpack, I basically took an existing vscode langpack and modified it by filling it with ADS resources and changing the package.json. This PR is for helping making those ADS resources easier to generate.

smartguest commented:
I just added the ability to be able to individually package a single extension to the build folder along with exporting the contents of the build folder as XLF. Currently the ADS-exclusive extensions will be packaged to the build folder and exported automatically, but the option  to create an XLF file or other extensions is available now.

smartguest commented:
I modified it so that the task now works on all extensions, following the template of the compile-extensions, also the packaging extensions task doesn't build extensions if they aren't webpacked, so I added the compilation step beforehand.

Charles-Gagnon commented:
You have a merge conflict to fix

Charles-Gagnon commented on build/lib/locFunc.ts:
```diff
@@ -24,33 +21,26 @@ export function packageLangpacksStream(): NodeJS.ReadWriteStream {
-
-			es.readArray(files).pipe(result);
+// Modified packageLocalExtensionsStream but for any ADS extensions including excluded/external ones.
+export function packageSingleADSExtensionStream(name : string): NodeJS.ReadWriteStream {
```
This doesn't really seem to have anything to do specifically with ADS extensions - could this just be named `packageSingleExtensionStream` or something to make it clearer?
